packages/components/src/panel/body.tsx

```diff
@@ -93,7 +101,16 @@ export function PanelBody(
 }
 
 const PanelBodyTitle = forwardRef(
-	( { isOpened, icon, title, ...props }, ref ) => {
```

Comment:
Thanks again! I really appreciate that you shared your thought process. Not only have I learned new things from your comment but also, it's really nice to see that we shared some of the same thoughts (even if I didn't have the context to execute them at the time). 

Anyways, this was all very clear! I'd just like to summarize my understanding of the below change to types, to ensure I am following:

```ts
buttonProps?: WordPressComponentProps< 
Omit< ButtonAsButtonProps, 'icon' >,
'button',
false>;
```
For our `buttonProps` prop, we are assigning the imported type `ButtonAsButtonProps` but omitting the icon prop since we need to specify a different type, `JSX.Element` in this case. By wrapping the above in the imported`WordPressComponentProps` type, it allows us to inherit props from `button` so we get props like `onClick` which would otherwise be missing. 

>But there's one more nuance: since `ButtonAsButtonProps` is not wrapped in `WordPressComponentProps` it doesn't include a bunch of props that PanelBodyTitle expects, like `onClick`

I'm curious, why is this the case?

Reply:
Also, do you mind sharing more context on the following change, with the utility type being on the right side of the operator? I'm not sure what this is called to look for it in the documentation. 

```ts
export type PanelBodyTitleProps = Omit< ButtonAsButtonProps, 'icon' > & {
``` 

It appears to be passing `ButtonAsButtonProps` (minus the icon) to the rest props. In that case, would it make sense to change the above to this instead: 

```ts
export type PanelBodyTitleProps = WordPressComponentProps<
	Omit< ButtonAsButtonProps, 'icon' >,
	'button',
	false
> & {
```
And remove the following from `packages/components/src/panel/body.tsx` so we don't import `WordPressComponentProps` in two different files? 

https://github.com/WordPress/gutenberg/blob/db1449e741312ad48d8c412767117cd00dafb834/packages/components/src/panel/body.tsx#L111

Reply:
> For our buttonProps prop, we are assigning the imported type ButtonAsButtonProps but omitting the icon prop since we need to specify a different type, JSX.Element in this case.

Correct!

> By wrapping the above in the importedWordPressComponentProps type, it allows us to inherit props from button so we get props like onClick which would otherwise be missing.

Yup :) 


> > But there's one more nuance: since `ButtonAsButtonProps` is not wrapped in `WordPressComponentProps` it doesn't include a bunch of props that PanelBodyTitle expects, like `onClick`
> 
> I'm curious, why is this the case?

`PanelBodyTitle` expects an `onClick` prop, but `ButtonAsButtonProps` doesn't have an `onClick`. For the `Button` component, that is not an issue because those [props are being wrapped in `WordPressComponentProps` when declaring `ButtonProps`](https://github.com/WordPress/gutenberg/blob/2ce55071deefcc8b386439e9bc71a8e3a9c8a6ea/packages/components/src/button/types.ts) — as you also explained above, that's where the `onClick` prop is added, together with every other prop that an HTML `button` element expects.

In our case, we need to add that `WordPressComponentsProps` wrap again so that we also get the `onClick` etc.

> Also, do you mind sharing more context on the following change, with the utility type being on the right side of the operator? 

I'm not sure exactly what you're referring to with "utility type", but I'll try my best to explain what's going on there:

The `&` operator allows to "put together" two different types (ie. `ResultingType = TypeA & TypeB`) — the resulting type is going to have all properties from `TypeA` and all propertied from `TypeB` together.
In this case, it means that `PanelBodyTitleProps` is going to have all props from `ButtonAsButtonProps` (minus `icon`), plus all props from that second type (ie. the `icon`, `isOpened` and `title` props).

> In that case, would it make sense to change the above to this instead:
> 
> ```ts
> export type PanelBodyTitleProps = WordPressComponentProps<
> 	Omit< ButtonAsButtonProps, 'icon' >,
> 	'button',
> 	false
> > & {
> ```
> 
> And remove the following from `packages/components/src/panel/body.tsx` so we don't import `WordPressComponentProps` in two different files?

Yes, that could be done — in this case it would be mostly a "stylistic" preference.

Personally, for example, I prefer to wrap prop types with `WordPressComponentProps` where the React component is declared, since in my mind it makes more sense to me that I know about all the extra props inherited by a component on top of its "own" props when looking at the component itself.

Again, hopefully I was clear enough! But as always feel free to ask as many questions as you need!

Reply:
I should have been more clear when asking 'why is this the case'. I was referring to: 

>ButtonAsButtonProps is not wrapped in WordPressComponentProps

So I was just curious about why `ButtonAsButtonProps` isn't wrapped in `WordPressComponentProps`. 

>I'm not sure exactly what you're referring to with "utility type"

I found [documentation on Omit here](https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys) under 'Utility Types', which is what I was referring to there.

Reply:
> So I was just curious about why ButtonAsButtonProps isn't wrapped in `WordPressComponentProps`.

It gets wrapped in `WordPressComponentProps` when it's used in `ButtonProps`:

https://github.com/WordPress/gutenberg/blob/2ce55071deefcc8b386439e9bc71a8e3a9c8a6ea/packages/components/src/button/types.ts#L13-L15

This ties a bit with what I mentioned about regarding the separation of a component's "own props" vs the props that a component inherits via `WordPressComponentProps`.

Having the `ButtonAsButtonProps` type not wrapped in `WordPressComponentProps` gives us more flexibility on how to use it.

Having said that, this is not an area where is only one correct approach — we also constantly iterate and re-assess our approaches, in case we find a new "pattern" that suits our needs better.

> I found [documentation on Omit here](https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys) under 'Utility Types', which is what I was referring to there.

Got it! Do you still have questions around that line of code?